{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c43f17df",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np\n",
    "import os\n",
    "from tqdm import tqdm\n",
    "from glob import glob\n",
    "import requests\n",
    "from bs4 import BeautifulSoup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4e90fbad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All genes in OncoKB: 1203\n",
      "Oncokb annotated genes: 950\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>hugoSymbol</th>\n",
       "      <th>entrezGeneId</th>\n",
       "      <th>grch37Isoform</th>\n",
       "      <th>grch37RefSeq</th>\n",
       "      <th>grch38Isoform</th>\n",
       "      <th>grch38RefSeq</th>\n",
       "      <th>oncokbAnnotated</th>\n",
       "      <th>occurrenceCount</th>\n",
       "      <th>mSKImpact</th>\n",
       "      <th>mSKHeme</th>\n",
       "      <th>foundation</th>\n",
       "      <th>foundationHeme</th>\n",
       "      <th>vogelstein</th>\n",
       "      <th>sangerCGC</th>\n",
       "      <th>geneType</th>\n",
       "      <th>geneAliases</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>25</td>\n",
       "      <td>ENST00000318560</td>\n",
       "      <td>NM_005157.4</td>\n",
       "      <td>ENST00000318560</td>\n",
       "      <td>NM_005157.4</td>\n",
       "      <td>True</td>\n",
       "      <td>7</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>ONCOGENE</td>\n",
       "      <td>[JTK7, c-ABL, ABL]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  hugoSymbol  entrezGeneId    grch37Isoform grch37RefSeq    grch38Isoform  \\\n",
       "0       ABL1            25  ENST00000318560  NM_005157.4  ENST00000318560   \n",
       "\n",
       "  grch38RefSeq  oncokbAnnotated  occurrenceCount  mSKImpact  mSKHeme  \\\n",
       "0  NM_005157.4             True                7       True     True   \n",
       "\n",
       "   foundation  foundationHeme  vogelstein  sangerCGC  geneType  \\\n",
       "0        True            True        True       True  ONCOGENE   \n",
       "\n",
       "          geneAliases  \n",
       "0  [JTK7, c-ABL, ABL]  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_genes = requests.get('https://www.oncokb.org/api/v1/utils/cancerGeneList', \n",
    "                         headers = {'Authorization': 'Bearer 948ece93-2979-4428-bef1-755d967f1a7a'}).json()\n",
    "all_genes = pd.DataFrame(all_genes)\n",
    "print('All genes in OncoKB:', all_genes.shape[0])\n",
    "all_genes = all_genes[(all_genes.oncokbAnnotated == True)]\n",
    "print('Oncokb annotated genes:', all_genes.shape[0])\n",
    "all_genes.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f0bab0ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "85 {'Other Biomarkers'}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Level</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Alterations</th>\n",
       "      <th>Cancer Types</th>\n",
       "      <th>Drugs (for therapeutic implications only)</th>\n",
       "      <th>alt_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>1</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>Colorectal Cancer</td>\n",
       "      <td>Cetuximab</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>1</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>Colorectal Cancer</td>\n",
       "      <td>Cetuximab + Chemotherapy</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>1</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>Colorectal Cancer</td>\n",
       "      <td>Panitumumab</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>1</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>Colorectal Cancer</td>\n",
       "      <td>Panitumumab + Chemotherapy</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>1</td>\n",
       "      <td>NRAS</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>Colorectal Cancer</td>\n",
       "      <td>Panitumumab</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>1</td>\n",
       "      <td>NRAS</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>Colorectal Cancer</td>\n",
       "      <td>Panitumumab + Chemotherapy</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>2</td>\n",
       "      <td>BRAF</td>\n",
       "      <td>v600 (excluding v600e and v600k)</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>Dabrafenib + Trametinib</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>2</td>\n",
       "      <td>BRAF</td>\n",
       "      <td>v600 (excluding v600e and v600k)</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>Encorafenib + Binimetinib</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>2</td>\n",
       "      <td>BRAF</td>\n",
       "      <td>v600 (excluding v600e and v600k)</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>Vemurafenib + Cobimetinib</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Level  Gene                       Alterations       Cancer Types  \\\n",
       "132     1  KRAS                          wildtype  Colorectal Cancer   \n",
       "133     1  KRAS                          wildtype  Colorectal Cancer   \n",
       "134     1  KRAS                          wildtype  Colorectal Cancer   \n",
       "135     1  KRAS                          wildtype  Colorectal Cancer   \n",
       "143     1  NRAS                          wildtype  Colorectal Cancer   \n",
       "144     1  NRAS                          wildtype  Colorectal Cancer   \n",
       "260     2  BRAF  v600 (excluding v600e and v600k)           Melanoma   \n",
       "261     2  BRAF  v600 (excluding v600e and v600k)           Melanoma   \n",
       "262     2  BRAF  v600 (excluding v600e and v600k)           Melanoma   \n",
       "\n",
       "    Drugs (for therapeutic implications only) alt_type  \n",
       "132                                 Cetuximab    other  \n",
       "133                  Cetuximab + Chemotherapy    other  \n",
       "134                               Panitumumab    other  \n",
       "135                Panitumumab + Chemotherapy    other  \n",
       "143                               Panitumumab    other  \n",
       "144                Panitumumab + Chemotherapy    other  \n",
       "260                   Dabrafenib + Trametinib    other  \n",
       "261                 Encorafenib + Binimetinib    other  \n",
       "262                 Vemurafenib + Cobimetinib    other  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#downloaded from the oncokb website - TODO: figure out a way to programatically download this\n",
    "table = pd.read_csv('raw-data/oncokb_biomarker_drug_associations.tsv', sep = '\\t')\n",
    "table = table[table['Drugs (for therapeutic implications only)'].notna()]\n",
    "def process_alterations(x):\n",
    "    x = x.lower()\n",
    "    if 'excluding' in x:\n",
    "        x = x\n",
    "    else:\n",
    "        if ' and ' in x:\n",
    "            x = x.split(' and ')\n",
    "        else:\n",
    "            x = x.split(',')\n",
    "    return x\n",
    "table['Alterations'] = table['Alterations'].apply(lambda x: process_alterations(x))\n",
    "table = table.explode('Alterations')\n",
    "table['Alterations'] = table['Alterations'].apply(lambda x: x.strip())\n",
    "def classify_alteration(x):\n",
    "    if any(i in x for i in ['fusion', 'translocation', 'rearrangement', 'insertion', 'inversion', 'duplication']):\n",
    "        return 'structural_variants'\n",
    "    if 'mutation' in x or 'insertion' in x:\n",
    "        return 'mutation'\n",
    "    if 'amplification' in x or 'deletion' in x or 'loss' in x or 'gain' in x:\n",
    "        return 'copy_number_alteration'\n",
    "    if 'microsatellite instability' in x:\n",
    "        return 'msi'\n",
    "    if len(x) == 4 or len(x) == 5 or len(x) == 6 or x[-3:] == 'del' or 'insfqea' in x or 'delinsd' in x: # e.g. E545K, V600E\n",
    "        return 'protein change'\n",
    "    else:\n",
    "        return 'other'\n",
    "table['alt_type'] = table['Alterations'].apply(classify_alteration)\n",
    "print(table['Gene'].nunique(), set(table['Gene']) - set(all_genes['hugoSymbol']))\n",
    "table[table.alt_type == 'other']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "147757a5",
   "metadata": {},
   "source": [
    "---\n",
    "### copy number alterations / mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "81e01df2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "65 65\n"
     ]
    }
   ],
   "source": [
    "cna = table[table.alt_type.isin(['copy_number_alteration', 'mutation'])][['Gene']].drop_duplicates()\n",
    "print(cna.Gene.nunique(), cna.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9a9cea05",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "82\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FGFR3</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KIT</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PIK3CA</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PDGFRA</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HRAS</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Other Biomarkers</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>RAD54L</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>RET</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MAP2K2</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MAP2K1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PTCH1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>IDH1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NRAS</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>FGFR1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ARAF</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>MTOR</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SF3B1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>SRSF2</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>BRAF</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>U2AF1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>RAF1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Gene  count\n",
       "0              FGFR3      4\n",
       "1                KIT      4\n",
       "2             PIK3CA      4\n",
       "3             PDGFRA      4\n",
       "4               HRAS      4\n",
       "5   Other Biomarkers      4\n",
       "6             RAD54L      4\n",
       "7                RET      4\n",
       "8             MAP2K2      4\n",
       "9             MAP2K1      4\n",
       "10             PTCH1      4\n",
       "11              KRAS      4\n",
       "12              IDH1      4\n",
       "13            ARID1A      4\n",
       "14              NRAS      4\n",
       "15              AKT1      4\n",
       "16             FGFR1      4\n",
       "17              ARAF      4\n",
       "18              MTOR      4\n",
       "19             SF3B1      4\n",
       "20             SRSF2      4\n",
       "21              BRAF      4\n",
       "22             U2AF1      4\n",
       "23              RAF1      4"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.makedirs('raw-data/cna', exist_ok=True)\n",
    "missing = []\n",
    "for i, row in (cna.iterrows()):\n",
    "    gene = row['Gene']\n",
    "    for t in ['GAIN', 'LOSS', 'AMPLIFICATION', 'DELETION']:\n",
    "        url = f'https://www.oncokb.org/api/v1/annotate/copyNumberAlterations?hugoSymbol={gene}&copyNameAlterationType={t}&referenceGenome=GRCh37'\n",
    "        api_out = requests.get(url, headers = {'Authorization': 'Bearer 948ece93-2979-4428-bef1-755d967f1a7a'}).json()['treatments']\n",
    "        if len(api_out) == 0:\n",
    "            missing.append([gene, t])\n",
    "            continue\n",
    "        df = pd.DataFrame(api_out)\n",
    "        df.to_csv(f'raw-data/cna/{gene}_{t}.csv', index=False)\n",
    "from glob import glob\n",
    "print(len(glob('raw-data/cna/*.csv')))\n",
    "missing = pd.DataFrame(missing, columns = ['Gene', 'Type'])\n",
    "missing = missing['Gene'].value_counts().reset_index(name = 'count')\n",
    "missing[missing['count'] == 4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "94f70e19",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "alt_type\n",
       "protein change            187\n",
       "mutation                  115\n",
       "structural_variants        20\n",
       "msi                        10\n",
       "other                       9\n",
       "copy_number_alteration      1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table[table['Gene'].isin(missing[missing['count'] == 4]['Gene'])].alt_type.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d2a3796a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Level</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Alterations</th>\n",
       "      <th>Cancer Types</th>\n",
       "      <th>Drugs (for therapeutic implications only)</th>\n",
       "      <th>alt_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>1</td>\n",
       "      <td>PDGFRA</td>\n",
       "      <td>exon 18 in-frame deletions</td>\n",
       "      <td>Gastrointestinal Stromal Tumor</td>\n",
       "      <td>Avapritinib</td>\n",
       "      <td>copy_number_alteration</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Level    Gene                 Alterations                    Cancer Types  \\\n",
       "166     1  PDGFRA  exon 18 in-frame deletions  Gastrointestinal Stromal Tumor   \n",
       "\n",
       "    Drugs (for therapeutic implications only)                alt_type  \n",
       "166                               Avapritinib  copy_number_alteration  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table[(table['alt_type'] == 'copy_number_alteration')&(table['Gene'].isin(missing[missing['count'] == 4]['Gene']))]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c048f64",
   "metadata": {},
   "source": [
    "---\n",
    "### fusions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3c458fb7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24 24\n"
     ]
    }
   ],
   "source": [
    "sv = table[table.alt_type.isin(['structural_variants'])][['Gene']].drop_duplicates()\n",
    "print(sv.Gene.nunique(), sv.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1fd8ad6f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERBB2</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FLT3</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PDGFB</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>RARA</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ESR1</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>FLI1</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene  count\n",
       "0   ABL1      7\n",
       "1   EGFR      7\n",
       "2  ERBB2      7\n",
       "3   FLT3      7\n",
       "4  PDGFB      7\n",
       "5   RARA      7\n",
       "6   ESR1      7\n",
       "7   FLI1      7"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.makedirs('raw-data/sv', exist_ok=True)\n",
    "missing = []\n",
    "for i, row in (sv.iterrows()):\n",
    "    gene = row['Gene']\n",
    "    df_all = pd.DataFrame()\n",
    "    for t in ['DELETION','TRANSLOCATION','DUPLICATION','INSERTION','INVERSION','FUSION','UNKNOWN']:\n",
    "        url = f'https://www.oncokb.org/api/v1/annotate/structuralVariants?hugoSymbolA={gene}&structuralVariantType={t}&referenceGenome=GRCh37&isFunctionalFusion=true'\n",
    "        api_out = requests.get(url, headers = {'Authorization': 'Bearer 948ece93-2979-4428-bef1-755d967f1a7a'}).json()['treatments']\n",
    "        df = pd.DataFrame(api_out)\n",
    "        df_all = pd.concat([df_all, df], axis = 0)\n",
    "        url = f'https://www.oncokb.org/api/v1/annotate/structuralVariants?hugoSymbolA={gene}&structuralVariantType={t}&referenceGenome=GRCh37&isFunctionalFusion=false'\n",
    "        api_out = requests.get(url, headers = {'Authorization': 'Bearer 948ece93-2979-4428-bef1-755d967f1a7a'}).json()['treatments']\n",
    "        df = pd.DataFrame(api_out)\n",
    "        df_all = pd.concat([df_all, df], axis = 0)\n",
    "        if df_all.shape[0] == 0:\n",
    "            missing.append([gene, t])\n",
    "    if df_all.shape[0] > 0:\n",
    "        df_all.to_csv(f'raw-data/sv/{gene}.csv', index=False)\n",
    "from glob import glob\n",
    "print(len(glob('raw-data/sv/*.csv')))\n",
    "missing = pd.DataFrame(missing, columns = ['Gene', 'Type'])\n",
    "missing = missing['Gene'].value_counts().reset_index(name = 'count')\n",
    "missing[missing['count'] == 7]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "4bb0618e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['bcr-abl1 fusion', 'col1a1-pdgfb fusion', 'pml-rara fusion', 'ewsr1-fli1 fusion']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene1</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Gene1, count]\n",
       "Index: []"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fusions = table[(table['Gene'].isin(missing[missing['count'] == 7]['Gene']))&\n",
    "      (table.alt_type == 'structural_variants')&\n",
    "      (table.Alterations.str.contains('fusion'))&\n",
    "      (table.Alterations.str.contains('-'))][['Gene', 'Alterations']].drop_duplicates()\n",
    "missing_fusions = []\n",
    "print(fusions.Alterations.tolist())\n",
    "for i, row in (fusions.iterrows()):\n",
    "    gene1 = row['Alterations'].split('-')[0]\n",
    "    gene2 = row['Alterations'].split('-')[1].split(' ')[0]\n",
    "    df_all = pd.DataFrame()\n",
    "    for t in ['DELETION','TRANSLOCATION','DUPLICATION','INSERTION','INVERSION','FUSION','UNKNOWN']:\n",
    "        url = f'https://www.oncokb.org/api/v1/annotate/structuralVariants?hugoSymbolA={gene1}&hugoSymbolB={gene2}&structuralVariantType={t}&referenceGenome=GRCh37&isFunctionalFusion=true'\n",
    "        api_out = requests.get(url, headers = {'Authorization': 'Bearer 948ece93-2979-4428-bef1-755d967f1a7a'}).json()['treatments']\n",
    "        df = pd.DataFrame(api_out)\n",
    "        df_all = pd.concat([df_all, df], axis = 0)\n",
    "        url = f'https://www.oncokb.org/api/v1/annotate/structuralVariants?hugoSymbolA={gene1}&hugoSymbolB={gene2}&structuralVariantType={t}&referenceGenome=GRCh37&isFunctionalFusion=false'\n",
    "        api_out = requests.get(url, headers = {'Authorization': 'Bearer 948ece93-2979-4428-bef1-755d967f1a7a'}).json()['treatments']\n",
    "        df = pd.DataFrame(api_out)\n",
    "        df_all = pd.concat([df_all, df], axis = 0)\n",
    "    if df_all.shape[0] == 0:\n",
    "        missing_fusions.append([gene1, gene2, t])\n",
    "    if df_all.shape[0] > 0:\n",
    "        df_all.to_csv(f'raw-data/sv/{gene1}-{gene2}.csv', index=False)\n",
    "missing_fusions = pd.DataFrame(missing_fusions, columns = ['Gene1', 'Gene2', 'Type'])\n",
    "missing_fusions = missing_fusions['Gene1'].value_counts().reset_index(name = 'count')\n",
    "missing_fusions[missing_fusions['count'] == 7]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a8f39977",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Alterations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>exon 20 in-frame insertions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>ERBB2</td>\n",
       "      <td>exon 20 in-frame insertions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>FLT3</td>\n",
       "      <td>internal tandem duplication</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>426</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>exon 19 in-frame insertions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>433</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>kinase domain duplication</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Gene                  Alterations\n",
       "75    EGFR  exon 20 in-frame insertions\n",
       "83   ERBB2  exon 20 in-frame insertions\n",
       "109   FLT3  internal tandem duplication\n",
       "426   EGFR  exon 19 in-frame insertions\n",
       "433   EGFR    kinase domain duplication"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table[(table['Gene'].isin(missing[missing['count'] == 7]['Gene']))&\n",
    "      (table.alt_type == 'structural_variants')&\n",
    "      (~table.Alterations.str.contains('fusion'))][['Gene', 'Alterations']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "336d672c",
   "metadata": {},
   "source": [
    "---\n",
    "### protein change"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "f951f11e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "239it [00:24,  9.62it/s]\n"
     ]
    }
   ],
   "source": [
    "protein_change = table[table.alt_type == 'protein change'][['Gene', 'Alterations']].drop_duplicates()\n",
    "os.makedirs('raw-data/protein_change', exist_ok=True)\n",
    "for i, row in tqdm(protein_change.iterrows()):\n",
    "    gene = row['Gene']\n",
    "    alteration = row['Alterations']\n",
    "    url = f'https://www.oncokb.org/api/v1/annotate/mutations/byProteinChange?referenceGenome=GRCh37&hugoSymbol={gene}&alteration={alteration}'\n",
    "    try:\n",
    "        api_out = requests.get(url, headers = {'Authorization': 'Bearer 948ece93-2979-4428-bef1-755d967f1a7a'}).json()['treatments']\n",
    "        df = pd.DataFrame(api_out)\n",
    "        df.to_csv(f'raw-data/protein_change/{gene}_{alteration}.csv', index=False)\n",
    "    except:\n",
    "        print(gene, alteration, 'error fetching, skipping')\n",
    "        continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acb604f2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e26ba274",
   "metadata": {},
   "source": [
    "----\n",
    "### other"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3a46283",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Alterations\n",
       "v600e                                16\n",
       "g12c                                 14\n",
       "t315i                                11\n",
       "microsatellite instability-high      10\n",
       "exon 19 in-frame deletions, l858r     9\n",
       "                                     ..\n",
       "a289v, r108k, t263p                   1\n",
       "l747p                                 1\n",
       "f1245y                                1\n",
       "c1156y                                1\n",
       "r1275q                                1\n",
       "Name: count, Length: 112, dtype: int64"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table = pd.read_csv('raw-data/oncokb_biomarker_drug_associations.tsv', sep = '\\t')\n",
    "table = table[table['Drugs (for therapeutic implications only)'].notna()]\n",
    "table['Alterations'] = table['Alterations'].apply(lambda x: x.lower())\n",
    "table[(table['Alterations'].str.contains('fusion') == False) & \n",
    "      (table['Alterations'].str.contains('amplification') == False) &\n",
    "      (table['Alterations'].str.contains('oncogenic mutations') == False)].Alterations.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cb6f679",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs('api-data', exist_ok=True)\n",
    "for i, row in tqdm(all_genes.iterrows()):\n",
    "    gene = row['hugoSymbol']\n",
    "    url = f'https://www.oncokb.org/api/v1/annotate/copyNumberAlterations?hugoSymbol={gene}&copyNameAlterationType=LOSS&referenceGenome=GRCh37'\n",
    "    api_out = requests.get(url, headers = {'Authorization': 'Bearer 948ece93-2979-4428-bef1-755d967f1a7a'}).json()['treatments']\n",
    "    if len(api_out) == 0:\n",
    "        continue\n",
    "    df = pd.DataFrame(api_out)\n",
    "    # processing - we'll do this later\n",
    "    # df['gene'] = gene\n",
    "    # df['alterations'] = df['alterations'].apply(lambda x: ','.join(x))\n",
    "    # df['drugs'] = df['drugs'].apply(lambda x: ','.join([i['drugName'] for i in x]))\n",
    "    # df['approvedIndications'] = df['approvedIndications'].apply(lambda x: x[0] if len(x) > 0 else np.nan)\n",
    "    # df['levelAssociatedCancerType'] = df['levelAssociatedCancerType'].apply(lambda x: x['name'] if x['name'] != '' else x['mainType']['name'])\n",
    "    # df['levelExcludedCancerTypes'] = df['levelExcludedCancerTypes'].apply(lambda x: ','.join(x))\n",
    "    # df['pmids'] = df['pmids'].apply(lambda x: ','.join(x))\n",
    "    # df['abstracts'] = df['abstracts'].apply(lambda x: ','.join([i['link'] for i in x]))\n",
    "    df.to_csv(f'api-data/{gene}.csv', index=False)\n",
    "from glob import glob\n",
    "len(glob('api-data/*.csv'))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
