{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6312fc17",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e342fbb0",
   "metadata": {},
   "source": [
    "---\n",
    "### sample info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "91c1b671",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PATIENT_ID</th>\n",
       "      <th>GENDER</th>\n",
       "      <th>RACE</th>\n",
       "      <th>ETHNICITY</th>\n",
       "      <th>CURRENT_AGE_DEID</th>\n",
       "      <th>STAGE_HIGHEST_RECORDED</th>\n",
       "      <th>NUM_ICDO_DX</th>\n",
       "      <th>ADRENAL_GLANDS</th>\n",
       "      <th>BONE</th>\n",
       "      <th>CNS_BRAIN</th>\n",
       "      <th>...</th>\n",
       "      <th>REPRODUCTIVE_ORGANS</th>\n",
       "      <th>SMOKING_PREDICTIONS_3_CLASSES</th>\n",
       "      <th>GLEASON_FIRST_REPORTED</th>\n",
       "      <th>GLEASON_HIGHEST_REPORTED</th>\n",
       "      <th>HISTORY_OF_PDL1</th>\n",
       "      <th>PRIOR_MED_TO_MSK</th>\n",
       "      <th>OS_MONTHS</th>\n",
       "      <th>OS_STATUS</th>\n",
       "      <th>HR</th>\n",
       "      <th>HER2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P-0000012</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Spanish; Non-Hispanic</td>\n",
       "      <td>68.0</td>\n",
       "      <td>Stage 1-3</td>\n",
       "      <td>2</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>Former/Current Smoker</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>118.454665</td>\n",
       "      <td>0:LIVING</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  PATIENT_ID  GENDER   RACE                  ETHNICITY  CURRENT_AGE_DEID  \\\n",
       "0  P-0000012  Female  White  Non-Spanish; Non-Hispanic              68.0   \n",
       "\n",
       "  STAGE_HIGHEST_RECORDED  NUM_ICDO_DX ADRENAL_GLANDS BONE CNS_BRAIN  ...  \\\n",
       "0              Stage 1-3            2             No   No        No  ...   \n",
       "\n",
       "  REPRODUCTIVE_ORGANS SMOKING_PREDICTIONS_3_CLASSES GLEASON_FIRST_REPORTED  \\\n",
       "0                  No         Former/Current Smoker                    NaN   \n",
       "\n",
       "  GLEASON_HIGHEST_REPORTED HISTORY_OF_PDL1 PRIOR_MED_TO_MSK   OS_MONTHS  \\\n",
       "0                      NaN              No          Unknown  118.454665   \n",
       "\n",
       "  OS_STATUS  HR  HER2  \n",
       "0  0:LIVING  No    No  \n",
       "\n",
       "[1 rows x 26 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "patient = pd.read_csv('msk_chord_2024/data_clinical_patient.txt', sep='\\t', comment='#')\n",
    "patient.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "7c0da9a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SAMPLE_ID</th>\n",
       "      <th>PATIENT_ID</th>\n",
       "      <th>GLEASON_SAMPLE_LEVEL</th>\n",
       "      <th>PDL1_POSITIVE</th>\n",
       "      <th>CANCER_TYPE</th>\n",
       "      <th>SAMPLE_TYPE</th>\n",
       "      <th>SAMPLE_CLASS</th>\n",
       "      <th>METASTATIC_SITE</th>\n",
       "      <th>PRIMARY_SITE</th>\n",
       "      <th>CANCER_TYPE_DETAILED</th>\n",
       "      <th>...</th>\n",
       "      <th>MSI_COMMENT</th>\n",
       "      <th>MSI_SCORE</th>\n",
       "      <th>MSI_TYPE</th>\n",
       "      <th>SOMATIC_STATUS</th>\n",
       "      <th>CLINICAL_GROUP</th>\n",
       "      <th>PATHOLOGICAL_GROUP</th>\n",
       "      <th>CLINICAL_SUMMARY</th>\n",
       "      <th>ICD_O_HISTOLOGY_DESCRIPTION</th>\n",
       "      <th>DIAGNOSIS_DESCRIPTION</th>\n",
       "      <th>TMB_NONSYNONYMOUS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P-0000012-T03-IM3</td>\n",
       "      <td>P-0000012</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>Metastasis</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Neck</td>\n",
       "      <td>Lung</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>...</td>\n",
       "      <td>MICROSATELLITE STABLE (MSS). See MSI note below.</td>\n",
       "      <td>0.47</td>\n",
       "      <td>Stable</td>\n",
       "      <td>Matched</td>\n",
       "      <td>3B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Distant</td>\n",
       "      <td>Adenocarcinoma, Nos</td>\n",
       "      <td>Lung and Bronchus</td>\n",
       "      <td>32.165504</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           SAMPLE_ID PATIENT_ID  GLEASON_SAMPLE_LEVEL PDL1_POSITIVE  \\\n",
       "0  P-0000012-T03-IM3  P-0000012                   NaN           NaN   \n",
       "\n",
       "                  CANCER_TYPE SAMPLE_TYPE SAMPLE_CLASS METASTATIC_SITE  \\\n",
       "0  Non-Small Cell Lung Cancer  Metastasis        Tumor            Neck   \n",
       "\n",
       "  PRIMARY_SITE CANCER_TYPE_DETAILED  ...  \\\n",
       "0         Lung  Lung Adenocarcinoma  ...   \n",
       "\n",
       "                                        MSI_COMMENT  MSI_SCORE  MSI_TYPE  \\\n",
       "0  MICROSATELLITE STABLE (MSS). See MSI note below.       0.47    Stable   \n",
       "\n",
       "  SOMATIC_STATUS CLINICAL_GROUP  PATHOLOGICAL_GROUP CLINICAL_SUMMARY  \\\n",
       "0        Matched             3B                 NaN          Distant   \n",
       "\n",
       "  ICD_O_HISTOLOGY_DESCRIPTION DIAGNOSIS_DESCRIPTION TMB_NONSYNONYMOUS  \n",
       "0         Adenocarcinoma, Nos     Lung and Bronchus         32.165504  \n",
       "\n",
       "[1 rows x 24 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample = pd.read_csv('msk_chord_2024/data_clinical_sample.txt', sep='\\t', comment='#')\n",
    "sample.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "b29a60d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PATIENT_ID</th>\n",
       "      <th>START_DATE</th>\n",
       "      <th>STOP_DATE</th>\n",
       "      <th>EVENT_TYPE</th>\n",
       "      <th>SUBTYPE</th>\n",
       "      <th>AGENT</th>\n",
       "      <th>RX_INVESTIGATIVE</th>\n",
       "      <th>FLAG_OROTOPICAL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P-0000012</td>\n",
       "      <td>-5437</td>\n",
       "      <td>-5369</td>\n",
       "      <td>Treatment</td>\n",
       "      <td>Chemo</td>\n",
       "      <td>CYCLOPHOSPHAMIDE</td>\n",
       "      <td>N</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  PATIENT_ID  START_DATE  STOP_DATE EVENT_TYPE SUBTYPE             AGENT  \\\n",
       "0  P-0000012       -5437      -5369  Treatment   Chemo  CYCLOPHOSPHAMIDE   \n",
       "\n",
       "  RX_INVESTIGATIVE  FLAG_OROTOPICAL  \n",
       "0                N                0  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "treatment = pd.read_csv('msk_chord_2024/data_timeline_treatment.txt', sep='\\t', comment='#')\n",
    "treatment.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "12a437e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PATIENT_ID\n",
       "P-0000012    2\n",
       "P-0021156    2\n",
       "P-0070191    2\n",
       "P-0044645    2\n",
       "P-0008662    2\n",
       "            ..\n",
       "P-0023104    1\n",
       "P-0023131    1\n",
       "P-0023173    1\n",
       "P-0023109    1\n",
       "P-0029444    1\n",
       "Name: count, Length: 24950, dtype: int64"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample['PATIENT_ID'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "2bd0f181",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PATIENT_ID</th>\n",
       "      <th>START_DATE</th>\n",
       "      <th>STOP_DATE</th>\n",
       "      <th>EVENT_TYPE</th>\n",
       "      <th>SUBTYPE</th>\n",
       "      <th>AGENT</th>\n",
       "      <th>RX_INVESTIGATIVE</th>\n",
       "      <th>FLAG_OROTOPICAL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>P-0000036</td>\n",
       "      <td>22</td>\n",
       "      <td>931</td>\n",
       "      <td>Treatment</td>\n",
       "      <td>Targeted</td>\n",
       "      <td>CRIZOTINIB</td>\n",
       "      <td>N</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>P-0000036</td>\n",
       "      <td>931</td>\n",
       "      <td>3512</td>\n",
       "      <td>Treatment</td>\n",
       "      <td>Targeted</td>\n",
       "      <td>CRIZOTINIB</td>\n",
       "      <td>N</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>P-0000058</td>\n",
       "      <td>1289</td>\n",
       "      <td>1629</td>\n",
       "      <td>Treatment</td>\n",
       "      <td>Targeted</td>\n",
       "      <td>LAPATINIB</td>\n",
       "      <td>N</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>P-0000058</td>\n",
       "      <td>1540</td>\n",
       "      <td>1733</td>\n",
       "      <td>Treatment</td>\n",
       "      <td>Targeted</td>\n",
       "      <td>ABEMACICLIB</td>\n",
       "      <td>N</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>P-0000066</td>\n",
       "      <td>629</td>\n",
       "      <td>1589</td>\n",
       "      <td>Treatment</td>\n",
       "      <td>Targeted</td>\n",
       "      <td>ABEMACICLIB</td>\n",
       "      <td>N</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134876</th>\n",
       "      <td>P-0089403</td>\n",
       "      <td>147</td>\n",
       "      <td>177</td>\n",
       "      <td>Treatment</td>\n",
       "      <td>Targeted</td>\n",
       "      <td>OLAPARIB</td>\n",
       "      <td>N</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134882</th>\n",
       "      <td>P-0089413</td>\n",
       "      <td>-400</td>\n",
       "      <td>225</td>\n",
       "      <td>Treatment</td>\n",
       "      <td>Targeted</td>\n",
       "      <td>ABEMACICLIB</td>\n",
       "      <td>N</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134888</th>\n",
       "      <td>P-0089445</td>\n",
       "      <td>-212</td>\n",
       "      <td>80</td>\n",
       "      <td>Treatment</td>\n",
       "      <td>Targeted</td>\n",
       "      <td>PALBOCICLIB</td>\n",
       "      <td>N</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134906</th>\n",
       "      <td>P-0089490</td>\n",
       "      <td>22</td>\n",
       "      <td>181</td>\n",
       "      <td>Treatment</td>\n",
       "      <td>Targeted</td>\n",
       "      <td>OSIMERTINIB</td>\n",
       "      <td>N</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134917</th>\n",
       "      <td>P-0089505</td>\n",
       "      <td>86</td>\n",
       "      <td>121</td>\n",
       "      <td>Treatment</td>\n",
       "      <td>Targeted</td>\n",
       "      <td>RIBOCICLIB</td>\n",
       "      <td>N</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9417 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       PATIENT_ID  START_DATE  STOP_DATE EVENT_TYPE   SUBTYPE        AGENT  \\\n",
       "30      P-0000036          22        931  Treatment  Targeted   CRIZOTINIB   \n",
       "31      P-0000036         931       3512  Treatment  Targeted   CRIZOTINIB   \n",
       "55      P-0000058        1289       1629  Treatment  Targeted    LAPATINIB   \n",
       "61      P-0000058        1540       1733  Treatment  Targeted  ABEMACICLIB   \n",
       "70      P-0000066         629       1589  Treatment  Targeted  ABEMACICLIB   \n",
       "...           ...         ...        ...        ...       ...          ...   \n",
       "134876  P-0089403         147        177  Treatment  Targeted     OLAPARIB   \n",
       "134882  P-0089413        -400        225  Treatment  Targeted  ABEMACICLIB   \n",
       "134888  P-0089445        -212         80  Treatment  Targeted  PALBOCICLIB   \n",
       "134906  P-0089490          22        181  Treatment  Targeted  OSIMERTINIB   \n",
       "134917  P-0089505          86        121  Treatment  Targeted   RIBOCICLIB   \n",
       "\n",
       "       RX_INVESTIGATIVE  FLAG_OROTOPICAL  \n",
       "30                    N                1  \n",
       "31                    N                1  \n",
       "55                    N                1  \n",
       "61                    N                1  \n",
       "70                    N                1  \n",
       "...                 ...              ...  \n",
       "134876                N                1  \n",
       "134882                N                1  \n",
       "134888                N                1  \n",
       "134906                N                1  \n",
       "134917                N                1  \n",
       "\n",
       "[9417 rows x 8 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "treatment[(treatment.SUBTYPE == 'Targeted')&(treatment['PATIENT_ID'].isin(sample['PATIENT_ID'].value_counts()[sample['PATIENT_ID'].value_counts() == 1].index))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "658d0bc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#choose a random sample with a single sample + has been treated with targeted therapy to test\n",
    "p = 'P-0000036'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "b6448343",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PATIENT_ID</th>\n",
       "      <th>START_DATE</th>\n",
       "      <th>STOP_DATE</th>\n",
       "      <th>EVENT_TYPE</th>\n",
       "      <th>SUBTYPE</th>\n",
       "      <th>AGENT</th>\n",
       "      <th>RX_INVESTIGATIVE</th>\n",
       "      <th>FLAG_OROTOPICAL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>P-0000036</td>\n",
       "      <td>-199</td>\n",
       "      <td>-31</td>\n",
       "      <td>Treatment</td>\n",
       "      <td>Biologic</td>\n",
       "      <td>BEVACIZUMAB</td>\n",
       "      <td>N</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>P-0000036</td>\n",
       "      <td>-199</td>\n",
       "      <td>-178</td>\n",
       "      <td>Treatment</td>\n",
       "      <td>Chemo</td>\n",
       "      <td>CARBOPLATIN</td>\n",
       "      <td>N</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>P-0000036</td>\n",
       "      <td>-199</td>\n",
       "      <td>-31</td>\n",
       "      <td>Treatment</td>\n",
       "      <td>Chemo</td>\n",
       "      <td>PEMETREXED</td>\n",
       "      <td>N</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>P-0000036</td>\n",
       "      <td>22</td>\n",
       "      <td>931</td>\n",
       "      <td>Treatment</td>\n",
       "      <td>Targeted</td>\n",
       "      <td>CRIZOTINIB</td>\n",
       "      <td>N</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>P-0000036</td>\n",
       "      <td>931</td>\n",
       "      <td>3512</td>\n",
       "      <td>Treatment</td>\n",
       "      <td>Targeted</td>\n",
       "      <td>CRIZOTINIB</td>\n",
       "      <td>N</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   PATIENT_ID  START_DATE  STOP_DATE EVENT_TYPE   SUBTYPE        AGENT  \\\n",
       "27  P-0000036        -199        -31  Treatment  Biologic  BEVACIZUMAB   \n",
       "28  P-0000036        -199       -178  Treatment     Chemo  CARBOPLATIN   \n",
       "29  P-0000036        -199        -31  Treatment     Chemo   PEMETREXED   \n",
       "30  P-0000036          22        931  Treatment  Targeted   CRIZOTINIB   \n",
       "31  P-0000036         931       3512  Treatment  Targeted   CRIZOTINIB   \n",
       "\n",
       "   RX_INVESTIGATIVE  FLAG_OROTOPICAL  \n",
       "27                N                0  \n",
       "28                N                0  \n",
       "29                N                0  \n",
       "30                N                1  \n",
       "31                N                1  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "treatment[treatment['PATIENT_ID'] == p]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ec85c7e",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "25ca3cc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PATIENT_ID                                       P-0000036\n",
       "GENDER                                              Female\n",
       "RACE                                                 Other\n",
       "ETHNICITY                        Non-Spanish; Non-Hispanic\n",
       "CURRENT_AGE_DEID                                      68.0\n",
       "STAGE_HIGHEST_RECORDED                             Stage 4\n",
       "NUM_ICDO_DX                                              1\n",
       "ADRENAL_GLANDS                                          No\n",
       "BONE                                                   Yes\n",
       "CNS_BRAIN                                               No\n",
       "INTRA_ABDOMINAL                                         No\n",
       "LIVER                                                  Yes\n",
       "LUNG                                                   Yes\n",
       "LYMPH_NODES                                            Yes\n",
       "OTHER                                                   No\n",
       "PLEURA                                                 Yes\n",
       "REPRODUCTIVE_ORGANS                                     No\n",
       "SMOKING_PREDICTIONS_3_CLASSES                        Never\n",
       "GLEASON_FIRST_REPORTED                                 NaN\n",
       "GLEASON_HIGHEST_REPORTED                               NaN\n",
       "HISTORY_OF_PDL1                                        NaN\n",
       "PRIOR_MED_TO_MSK                                   Unknown\n",
       "OS_MONTHS                                       115.462887\n",
       "OS_STATUS                                         0:LIVING\n",
       "HR                                                     NaN\n",
       "HER2                                                   NaN\n",
       "Name: 2, dtype: object"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "patient[patient['PATIENT_ID'] == p].iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "320041a1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SAMPLE_ID                               P-0000036-T01-IM3\n",
       "PATIENT_ID                                      P-0000036\n",
       "GLEASON_SAMPLE_LEVEL                                  NaN\n",
       "PDL1_POSITIVE                                         NaN\n",
       "CANCER_TYPE                    Non-Small Cell Lung Cancer\n",
       "SAMPLE_TYPE                                       Primary\n",
       "SAMPLE_CLASS                                        Tumor\n",
       "METASTATIC_SITE                            Not Applicable\n",
       "PRIMARY_SITE                                         Lung\n",
       "CANCER_TYPE_DETAILED                  Lung Adenocarcinoma\n",
       "GENE_PANEL                                      IMPACT341\n",
       "SAMPLE_COVERAGE                                       380\n",
       "TUMOR_PURITY                                         30.0\n",
       "ONCOTREE_CODE                                        LUAD\n",
       "MSI_COMMENT                                           NaN\n",
       "MSI_SCORE                                            -1.0\n",
       "MSI_TYPE                                    Do not report\n",
       "SOMATIC_STATUS                                  Unmatched\n",
       "CLINICAL_GROUP                                          4\n",
       "PATHOLOGICAL_GROUP                                    NaN\n",
       "CLINICAL_SUMMARY                                  Distant\n",
       "ICD_O_HISTOLOGY_DESCRIPTION           Adenocarcinoma, Nos\n",
       "DIAGNOSIS_DESCRIPTION                   Lung and Bronchus\n",
       "TMB_NONSYNONYMOUS                                7.764087\n",
       "Name: 3, dtype: object"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample[sample['PATIENT_ID'] == p].iloc[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca421f01",
   "metadata": {},
   "source": [
    "---\n",
    "### sample format "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "7d7ec8b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Patient ID: P-0000036\n",
      "Gender: Female\n",
      "Age: 68.0\n",
      "Race: Other\n",
      "Ethnicity: Non-Spanish; Non-Hispanic\n",
      "Stage (highest recorded): Stage 4\n",
      "Overall Survival: 115.5 months (0:LIVING)\n",
      "Prior Medication: Unknown\n",
      "\n",
      "Sample ID: P-0000036-T01-IM3\n",
      "Cancer Type: Lung Adenocarcinoma\n",
      "Primary Site: Lung\n",
      "Sample Type: Primary\n",
      "Tumor Purity: 30.0%\n",
      "Gene Panel: IMPACT341\n",
      "Coverage: 380\n",
      "TMB (nonsynonymous): 7.764087104\n",
      "MSI Status: Do not report (Score: -1.0)\n",
      "Somatic Status: Unmatched\n"
     ]
    }
   ],
   "source": [
    "def format_oncopanel_report(patient_df, sample_df):\n",
    "    # Convert single-row dataframes to dictionaries for easy access\n",
    "    patient_info = patient_df.to_dict()\n",
    "    sample_info = sample_df.to_dict()\n",
    "\n",
    "    # Build report text\n",
    "    report = []\n",
    "    report.append(f\"Patient ID: {patient_info.get('PATIENT_ID', 'N/A')}\")\n",
    "    report.append(f\"Gender: {patient_info.get('GENDER', 'N/A')}\")\n",
    "    report.append(f\"Age: {patient_info.get('CURRENT_AGE_DEID', 'N/A')}\")\n",
    "    report.append(f\"Race: {patient_info.get('RACE', 'N/A')}\")\n",
    "    report.append(f\"Ethnicity: {patient_info.get('ETHNICITY', 'N/A')}\")\n",
    "    report.append(f\"Stage (highest recorded): {patient_info.get('STAGE_HIGHEST_RECORDED', 'N/A')}\")\n",
    "    report.append(f\"Overall Survival: {patient_info.get('OS_MONTHS', 'N/A'):.1f} months ({patient_info.get('OS_STATUS', 'N/A')})\")\n",
    "    report.append(f\"Prior Medication: {patient_info.get('PRIOR_MED_TO_MSK', 'N/A')}\")\n",
    "\n",
    "    # Sample-specific details\n",
    "    report.append(f\"\\nSample ID: {sample_info.get('SAMPLE_ID', 'N/A')}\")\n",
    "    report.append(f\"Cancer Type: {sample_info.get('CANCER_TYPE_DETAILED', sample_info.get('CANCER_TYPE', 'N/A'))}\")\n",
    "    report.append(f\"Primary Site: {sample_info.get('PRIMARY_SITE', 'N/A')}\")\n",
    "    report.append(f\"Sample Type: {sample_info.get('SAMPLE_TYPE', 'N/A')}\")\n",
    "    report.append(f\"Tumor Purity: {sample_info.get('TUMOR_PURITY', 'N/A')}%\")\n",
    "    report.append(f\"Gene Panel: {sample_info.get('GENE_PANEL', 'N/A')}\")\n",
    "    report.append(f\"Coverage: {sample_info.get('SAMPLE_COVERAGE', 'N/A')}\")\n",
    "    report.append(f\"TMB (nonsynonymous): {sample_info.get('TMB_NONSYNONYMOUS', 'N/A')}\")\n",
    "    report.append(f\"MSI Status: {sample_info.get('MSI_TYPE', 'N/A')} (Score: {sample_info.get('MSI_SCORE', 'N/A')})\")\n",
    "    report.append(f\"Somatic Status: {sample_info.get('SOMATIC_STATUS', 'N/A')}\")\n",
    "\n",
    "    return \"\\n\".join(report)\n",
    "\n",
    "\n",
    "# Example usage:\n",
    "patient_df = patient[patient['PATIENT_ID'] == p].iloc[0]\n",
    "sample_df = sample[sample['PATIENT_ID'] == p].iloc[0]\n",
    "report_text = format_oncopanel_report(patient_df, sample_df)\n",
    "print(report_text)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "254819ba",
   "metadata": {},
   "source": [
    "----\n",
    "### sequencing info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2cb2e33",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>P-0008840-T01-IM5</th>\n",
       "      <th>P-0050951-T01-IM6</th>\n",
       "      <th>P-0086178-T01-IM7</th>\n",
       "      <th>P-0020358-T01-IM6</th>\n",
       "      <th>P-0089413-T01-IM7</th>\n",
       "      <th>P-0033156-T01-IM6</th>\n",
       "      <th>P-0044605-T01-IM6</th>\n",
       "      <th>P-0077282-T01-IM7</th>\n",
       "      <th>P-0037126-T01-IM6</th>\n",
       "      <th>...</th>\n",
       "      <th>P-0011072-T01-IM5</th>\n",
       "      <th>P-0072579-T02-IM7</th>\n",
       "      <th>P-0050748-T01-IM6</th>\n",
       "      <th>P-0087758-T01-IM7</th>\n",
       "      <th>P-0062922-T01-IM7</th>\n",
       "      <th>P-0052296-T02-IM6</th>\n",
       "      <th>P-0048082-T01-IM6</th>\n",
       "      <th>P-0069255-T01-IM7</th>\n",
       "      <th>P-0060577-T01-IH3</th>\n",
       "      <th>P-0070319-T01-IH3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TAP1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 25035 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol  P-0008840-T01-IM5  P-0050951-T01-IM6  P-0086178-T01-IM7  \\\n",
       "0        TAP1                0.0                0.0                0.0   \n",
       "\n",
       "   P-0020358-T01-IM6  P-0089413-T01-IM7  P-0033156-T01-IM6  P-0044605-T01-IM6  \\\n",
       "0                0.0                0.0                0.0                0.0   \n",
       "\n",
       "   P-0077282-T01-IM7  P-0037126-T01-IM6  ...  P-0011072-T01-IM5  \\\n",
       "0                0.0                0.0  ...                0.0   \n",
       "\n",
       "   P-0072579-T02-IM7  P-0050748-T01-IM6  P-0087758-T01-IM7  P-0062922-T01-IM7  \\\n",
       "0                0.0                0.0                0.0                0.0   \n",
       "\n",
       "   P-0052296-T02-IM6  P-0048082-T01-IM6  P-0069255-T01-IM7  P-0060577-T01-IH3  \\\n",
       "0                0.0                0.0                0.0                NaN   \n",
       "\n",
       "   P-0070319-T01-IH3  \n",
       "0                NaN  \n",
       "\n",
       "[1 rows x 25035 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cna = pd.read_csv('msk_chord_2024/data_cna.txt', sep='\\t', comment='#')\n",
    "cna.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "7d42b61b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>P-0000036-T01-IM3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Hugo_Symbol, P-0000036-T01-IM3]\n",
       "Index: []"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cna_sample = cna[['Hugo_Symbol', sample[sample['PATIENT_ID'] == p].iloc[0]['SAMPLE_ID']]]\n",
    "cna_sample[(cna_sample[sample[sample['PATIENT_ID'] == p].iloc[0]['SAMPLE_ID']] != 0) & (cna_sample[sample[sample['PATIENT_ID'] == p].iloc[0]['SAMPLE_ID']].notna())]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c116939c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/n_/hh_by1z91yd4l5wmkyclq4r00000gn/T/ipykernel_14857/1943742257.py:1: DtypeWarning: Columns (44,49,88) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  mut = pd.read_csv('msk_chord_2024/data_mutations.txt', sep='\\t', comment='#')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>Center</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>...</th>\n",
       "      <th>amino_acid_change</th>\n",
       "      <th>cDNA_Change</th>\n",
       "      <th>cDNA_position</th>\n",
       "      <th>cdna_change</th>\n",
       "      <th>comments</th>\n",
       "      <th>n_depth</th>\n",
       "      <th>t_depth</th>\n",
       "      <th>transcript</th>\n",
       "      <th>allelic_fraction</th>\n",
       "      <th>HGV</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>1956</td>\n",
       "      <td>MSKCC</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>7</td>\n",
       "      <td>55242470</td>\n",
       "      <td>55242487</td>\n",
       "      <td>+</td>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.474465</td>\n",
       "      <td>c.2240_2257del</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 125 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol  Entrez_Gene_Id Center NCBI_Build Chromosome  Start_Position  \\\n",
       "0        EGFR            1956  MSKCC     GRCh37          7        55242470   \n",
       "\n",
       "   End_Position Strand       Consequence Variant_Classification  ...  \\\n",
       "0      55242487      +  inframe_deletion           In_Frame_Del  ...   \n",
       "\n",
       "  amino_acid_change cDNA_Change cDNA_position cdna_change comments  n_depth  \\\n",
       "0               NaN         NaN           NaN         NaN      NaN      NaN   \n",
       "\n",
       "  t_depth  transcript  allelic_fraction             HGV  \n",
       "0     NaN         NaN          0.474465  c.2240_2257del  \n",
       "\n",
       "[1 rows x 125 columns]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut = pd.read_csv('msk_chord_2024/data_mutations.txt', sep='\\t', comment='#')\n",
    "mut['allelic_fraction'] = mut['t_alt_count'] / (mut['t_ref_count'] + mut['t_alt_count'])\n",
    "mut['HGV'] = mut['HGVSc'].apply(lambda x: x.split(':')[1] if pd.notna(x) and ':' in x else x)\n",
    "mut.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "37391756",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Hugo_Symbol', 'Entrez_Gene_Id', 'Center', 'NCBI_Build', 'Chromosome',\n",
       "       'Start_Position', 'End_Position', 'Strand', 'Consequence',\n",
       "       'Variant_Classification', 'Variant_Type', 'Reference_Allele',\n",
       "       'Tumor_Seq_Allele1', 'Tumor_Seq_Allele2', 'dbSNP_RS',\n",
       "       'Tumor_Sample_Barcode', 'Validation_Status', 'Mutation_Status', 'Score',\n",
       "       't_ref_count', 't_alt_count', 'n_ref_count', 'n_alt_count', 'HGVSc',\n",
       "       'HGVSp', 'HGVSp_Short', 'Transcript_ID', 'RefSeq', 'Protein_position',\n",
       "       'Codons', 'allelic_fraction', 'HGV'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut[mut['Tumor_Sample_Barcode'] == sample[sample['PATIENT_ID'] == p].iloc[0]['SAMPLE_ID']].dropna(axis=1, how='all').columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "69acfc4b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele1</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>dbSNP_RS</th>\n",
       "      <th>Validation_Status</th>\n",
       "      <th>Mutation_Status</th>\n",
       "      <th>allelic_fraction</th>\n",
       "      <th>HGVSp_Short</th>\n",
       "      <th>HGV</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>207948</th>\n",
       "      <td>NOTCH4</td>\n",
       "      <td>6</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>rs150079294</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>UNKNOWN</td>\n",
       "      <td>0.483108</td>\n",
       "      <td>p.C815G</td>\n",
       "      <td>c.2443T&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207949</th>\n",
       "      <td>IRS1</td>\n",
       "      <td>2</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>rs138145752</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>UNKNOWN</td>\n",
       "      <td>0.444976</td>\n",
       "      <td>p.E461G</td>\n",
       "      <td>c.1382A&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207950</th>\n",
       "      <td>TSHR</td>\n",
       "      <td>14</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>rs142063461</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>UNKNOWN</td>\n",
       "      <td>0.498062</td>\n",
       "      <td>p.P68S</td>\n",
       "      <td>c.202C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207951</th>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>stop_gained,splice_region_variant</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>UNKNOWN</td>\n",
       "      <td>0.057878</td>\n",
       "      <td>p.Q331*</td>\n",
       "      <td>c.991C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207952</th>\n",
       "      <td>ERBB2</td>\n",
       "      <td>17</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>rs140272156</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>UNKNOWN</td>\n",
       "      <td>0.420904</td>\n",
       "      <td>p.G1015E</td>\n",
       "      <td>c.3044G&gt;A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207953</th>\n",
       "      <td>AR</td>\n",
       "      <td>X</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>UNKNOWN</td>\n",
       "      <td>0.135274</td>\n",
       "      <td>p.L226V</td>\n",
       "      <td>c.676T&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207954</th>\n",
       "      <td>FBXW7</td>\n",
       "      <td>4</td>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>TCCTCCTCA</td>\n",
       "      <td>TCCTCCTCA</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>UNKNOWN</td>\n",
       "      <td>0.396825</td>\n",
       "      <td>p.D112_E114del</td>\n",
       "      <td>c.336_344del</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Hugo_Symbol Chromosome                        Consequence  \\\n",
       "207948      NOTCH4          6                   missense_variant   \n",
       "207949        IRS1          2                   missense_variant   \n",
       "207950        TSHR         14                   missense_variant   \n",
       "207951        TP53         17  stop_gained,splice_region_variant   \n",
       "207952       ERBB2         17                   missense_variant   \n",
       "207953          AR          X                   missense_variant   \n",
       "207954       FBXW7          4                   inframe_deletion   \n",
       "\n",
       "       Variant_Classification Variant_Type Reference_Allele Tumor_Seq_Allele1  \\\n",
       "207948      Missense_Mutation          SNP                A                 A   \n",
       "207949      Missense_Mutation          SNP                T                 T   \n",
       "207950      Missense_Mutation          SNP                C                 C   \n",
       "207951      Nonsense_Mutation          SNP                G                 G   \n",
       "207952      Missense_Mutation          SNP                G                 G   \n",
       "207953      Missense_Mutation          SNP                T                 T   \n",
       "207954           In_Frame_Del          DEL        TCCTCCTCA         TCCTCCTCA   \n",
       "\n",
       "       Tumor_Seq_Allele2     dbSNP_RS Validation_Status Mutation_Status  \\\n",
       "207948                 C  rs150079294           Unknown         UNKNOWN   \n",
       "207949                 C  rs138145752           Unknown         UNKNOWN   \n",
       "207950                 T  rs142063461           Unknown         UNKNOWN   \n",
       "207951                 A          NaN           Unknown         UNKNOWN   \n",
       "207952                 A  rs140272156           Unknown         UNKNOWN   \n",
       "207953                 G          NaN           Unknown         UNKNOWN   \n",
       "207954                 -          NaN           Unknown         UNKNOWN   \n",
       "\n",
       "        allelic_fraction     HGVSp_Short           HGV  \n",
       "207948          0.483108         p.C815G     c.2443T>G  \n",
       "207949          0.444976         p.E461G     c.1382A>G  \n",
       "207950          0.498062          p.P68S      c.202C>T  \n",
       "207951          0.057878         p.Q331*      c.991C>T  \n",
       "207952          0.420904        p.G1015E     c.3044G>A  \n",
       "207953          0.135274         p.L226V      c.676T>G  \n",
       "207954          0.396825  p.D112_E114del  c.336_344del  "
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut[mut['Tumor_Sample_Barcode'] == sample[sample['PATIENT_ID'] == p].iloc[0]['SAMPLE_ID']][['Hugo_Symbol', 'Chromosome', 'Consequence', \n",
    "                                                                                            'Variant_Classification', 'Variant_Type', 'Reference_Allele',\n",
    "       'Tumor_Seq_Allele1', 'Tumor_Seq_Allele2', 'dbSNP_RS',\n",
    "       'Validation_Status', 'Mutation_Status', 'allelic_fraction', 'HGVSp_Short', 'HGV']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "f5899e20",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_Id</th>\n",
       "      <th>SV_Status</th>\n",
       "      <th>Site1_Hugo_Symbol</th>\n",
       "      <th>Site2_Hugo_Symbol</th>\n",
       "      <th>Site1_Chromosome</th>\n",
       "      <th>Site2_Chromosome</th>\n",
       "      <th>Site1_Position</th>\n",
       "      <th>Site2_Position</th>\n",
       "      <th>Site1_Description</th>\n",
       "      <th>Site2_Description</th>\n",
       "      <th>...</th>\n",
       "      <th>Connection_Type</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>DNA_Support</th>\n",
       "      <th>RNA_Support</th>\n",
       "      <th>SV_Length</th>\n",
       "      <th>Normal_Read_Count</th>\n",
       "      <th>Tumor_Read_Count</th>\n",
       "      <th>Normal_Variant_Count</th>\n",
       "      <th>Tumor_Variant_Count</th>\n",
       "      <th>Comments</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P-0022424-T01-IM6</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>SEPTIN12</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>4838352.0</td>\n",
       "      <td>27089562.0</td>\n",
       "      <td>5-UTR of SEPT12(-):11Kb before coding start</td>\n",
       "      <td>Exon 8 of ARID1A(+)</td>\n",
       "      <td>...</td>\n",
       "      <td>3to3</td>\n",
       "      <td>ARID1A (NM_006015) - SEPT12 (NM_144605) rearra...</td>\n",
       "      <td>yes</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>Note: The ARID1A - SEPT12 rearrangement is a t...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Sample_Id SV_Status Site1_Hugo_Symbol Site2_Hugo_Symbol  \\\n",
       "0  P-0022424-T01-IM6   SOMATIC          SEPTIN12            ARID1A   \n",
       "\n",
       "  Site1_Chromosome Site2_Chromosome  Site1_Position  Site2_Position  \\\n",
       "0               16                1       4838352.0      27089562.0   \n",
       "\n",
       "                             Site1_Description    Site2_Description  ...  \\\n",
       "0  5-UTR of SEPT12(-):11Kb before coding start  Exon 8 of ARID1A(+)  ...   \n",
       "\n",
       "   Connection_Type                                         Annotation  \\\n",
       "0             3to3  ARID1A (NM_006015) - SEPT12 (NM_144605) rearra...   \n",
       "\n",
       "  DNA_Support  RNA_Support  SV_Length Normal_Read_Count Tumor_Read_Count  \\\n",
       "0         yes      unknown        0.0               0.0              0.0   \n",
       "\n",
       "  Normal_Variant_Count Tumor_Variant_Count  \\\n",
       "0                  0.0                10.0   \n",
       "\n",
       "                                            Comments  \n",
       "0  Note: The ARID1A - SEPT12 rearrangement is a t...  \n",
       "\n",
       "[1 rows x 27 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sv = pd.read_csv('msk_chord_2024/data_sv.txt', sep='\\t', comment='#')\n",
    "sv.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "eadbffca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Sample_Id                                              P-0000036-T01-IM3\n",
       "SV_Status                                                        SOMATIC\n",
       "Site1_Hugo_Symbol                                                   ROS1\n",
       "Site2_Hugo_Symbol                                                SLC34A2\n",
       "Site1_Chromosome                                                       6\n",
       "Site2_Chromosome                                                       4\n",
       "Site1_Position                                               117662374.0\n",
       "Site2_Position                                                25667150.0\n",
       "Site1_Description                                     Exon 30 of ROS1(-)\n",
       "Site2_Description              Intron of SLC34A2(+): 600bp before exon 5\n",
       "Site2_Effect_On_Frame                                                NaN\n",
       "NCBI_Build                                                        GRCh37\n",
       "Class                                                      TRANSLOCATION\n",
       "Tumor_Split_Read_Count                                              17.0\n",
       "Tumor_Paired_End_Read_Count                                          6.0\n",
       "Event_Info                                              Antisense fusion\n",
       "Breakpoint_Type                                     FFPE-PooledNormal_bc\n",
       "Connection_Type                                                     5to3\n",
       "Annotation                                                           NaN\n",
       "DNA_Support                                                          yes\n",
       "RNA_Support                                                      unknown\n",
       "SV_Length                                                            0.0\n",
       "Normal_Read_Count                                                    0.0\n",
       "Tumor_Read_Count                                                     0.0\n",
       "Normal_Variant_Count                                                 0.0\n",
       "Tumor_Variant_Count                                                  0.0\n",
       "Comments                                                             NaN\n",
       "Name: 4968, dtype: object"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sv[sv['Sample_Id'] == sample[sample['PATIENT_ID'] == p].iloc[0]['SAMPLE_ID']].iloc[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc7dc37a",
   "metadata": {},
   "source": [
    "### sequencing format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "91d30518",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'SNV/Indel':           Gene       Variant Type                        Consequence  \\\n",
       " 207953      AR  Missense_Mutation                   missense_variant   \n",
       " 207952   ERBB2  Missense_Mutation                   missense_variant   \n",
       " 207954   FBXW7       In_Frame_Del                   inframe_deletion   \n",
       " 207949    IRS1  Missense_Mutation                   missense_variant   \n",
       " 207948  NOTCH4  Missense_Mutation                   missense_variant   \n",
       " 207951    TP53  Nonsense_Mutation  stop_gained,splice_region_variant   \n",
       " 207950    TSHR  Missense_Mutation                   missense_variant   \n",
       " \n",
       "         Protein Change   cDNA Change  Allelic Fraction  \n",
       " 207953         p.L226V      c.676T>G             0.135  \n",
       " 207952        p.G1015E     c.3044G>A             0.421  \n",
       " 207954  p.D112_E114del  c.336_344del             0.397  \n",
       " 207949         p.E461G     c.1382A>G             0.445  \n",
       " 207948         p.C815G     c.2443T>G             0.483  \n",
       " 207951         p.Q331*      c.991C>T             0.058  \n",
       " 207950          p.P68S      c.202C>T             0.498  ,\n",
       " 'SV': Site1_Hugo_Symbol                          ROS1\n",
       " Site2_Hugo_Symbol                       SLC34A2\n",
       " Class                             TRANSLOCATION\n",
       " Event_Info                     Antisense fusion\n",
       " Tumor_Split_Read_Count                     17.0\n",
       " Tumor_Paired_End_Read_Count                 6.0\n",
       " DNA_Support                                 yes\n",
       " RNA_Support                             unknown\n",
       " Name: 4968, dtype: object}"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "def make_msk_variant_tables(snv_df=None, cna_df=None, sv_df=None, sample_id=None):\n",
    "    \"\"\"\n",
    "    Generate MSK-style mutation tables for SNV/indels, CNAs, and SVs.\n",
    "    Each is returned as a pandas DataFrame.\n",
    "    \"\"\"\n",
    "\n",
    "    tables = {}\n",
    "\n",
    "    # SNV/Indel table (if provided)\n",
    "    if snv_df is not None and not snv_df.empty:\n",
    "        snv_table = snv_df[\n",
    "            [\"Hugo_Symbol\", \"Variant_Classification\", \"Consequence\", \n",
    "             \"HGVSp_Short\", \"HGV\", \"allelic_fraction\"]\n",
    "        ].copy()\n",
    "        snv_table.columns = [\"Gene\", \"Variant Type\", \"Consequence\",\n",
    "                             \"Protein Change\", \"cDNA Change\", \"Allelic Fraction\"]\n",
    "        snv_table[\"Allelic Fraction\"] = snv_table[\"Allelic Fraction\"].round(3)\n",
    "        snv_table.sort_values([\"Gene\"], inplace=True)\n",
    "        tables[\"SNV/Indel\"] = snv_table\n",
    "\n",
    "    # CNA table\n",
    "    if cna_df is not None and sample_id in cna_df.columns and not cna_df.empty:\n",
    "        cna_table = cna_df[[\"Hugo_Symbol\", sample_id]].copy()\n",
    "        cna_table.columns = [\"Gene\", \"Copy Number Alteration\"]\n",
    "        cna_table = cna_table[cna_table[\"Copy Number Alteration\"] != 0]  # Remove neutral calls\n",
    "        tables[\"CNA\"] = cna_table\n",
    "\n",
    "    # SV / Fusion table\n",
    "    if sv_df is not None:\n",
    "        sv_table = sv_df[\n",
    "            [\"Site1_Hugo_Symbol\", \"Site2_Hugo_Symbol\", \"Class\", \"Event_Info\",\n",
    "             \"Tumor_Split_Read_Count\", \"Tumor_Paired_End_Read_Count\", \n",
    "             \"DNA_Support\", \"RNA_Support\"]\n",
    "        ].copy()\n",
    "        sv_table.columns = [\"Gene 1\", \"Gene 2\", \"Event Type\", \"Event Info\",\n",
    "                            \"Split Reads\", \"Paired-End Reads\", \"DNA Support\", \"RNA Support\"]\n",
    "        tables[\"SV\"] = sv_table\n",
    "\n",
    "    return tables\n",
    "\n",
    "\n",
    "# Example usage with your CNA and SV data\n",
    "# Assuming cna_df and sv_df are loaded pandas DataFrames, and sample_id is the sample name string\n",
    "\n",
    "snv_df = mut[mut['Tumor_Sample_Barcode'] == sample[sample['PATIENT_ID'] == p].iloc[0]['SAMPLE_ID']][['Hugo_Symbol', 'Chromosome', 'Consequence', \n",
    "                                                                                            'Variant_Classification', 'Variant_Type', 'Reference_Allele',\n",
    "       'Tumor_Seq_Allele1', 'Tumor_Seq_Allele2', 'dbSNP_RS',\n",
    "       'Validation_Status', 'Mutation_Status', 'allelic_fraction', 'HGVSp_Short', 'HGV']]\n",
    "cna_df = cna_sample[(cna_sample[sample[sample['PATIENT_ID'] == p].iloc[0]['SAMPLE_ID']] != 0) & (cna_sample[sample[sample['PATIENT_ID'] == p].iloc[0]['SAMPLE_ID']].notna())]\n",
    "sv_df = sv[sv['Sample_Id'] == sample[sample['PATIENT_ID'] == p].iloc[0]['SAMPLE_ID']].iloc[0]\n",
    "\n",
    "tables = make_msk_variant_tables(cna_df=cna_df, sv_df=sv_df, snv_df=snv_df, sample_id=\"P-0000036-T01-IM3\")\n",
    "\n",
    "tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "3740e16a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "DETECTED VARIANTS\n",
      "\n",
      "  Single Nucleotide Variants / Indels\n",
      "  Gene      Variant Type                       Consequence Protein Change  cDNA Change  Allelic Fraction\n",
      "    AR Missense_Mutation                  missense_variant        p.L226V     c.676T>G             0.135\n",
      " ERBB2 Missense_Mutation                  missense_variant       p.G1015E    c.3044G>A             0.421\n",
      " FBXW7      In_Frame_Del                  inframe_deletion p.D112_E114del c.336_344del             0.397\n",
      "  IRS1 Missense_Mutation                  missense_variant        p.E461G    c.1382A>G             0.445\n",
      "NOTCH4 Missense_Mutation                  missense_variant        p.C815G    c.2443T>G             0.483\n",
      "  TP53 Nonsense_Mutation stop_gained,splice_region_variant        p.Q331*     c.991C>T             0.058\n",
      "  TSHR Missense_Mutation                  missense_variant         p.P68S     c.202C>T             0.498\n",
      "\n",
      "  Structural Variants / Fusions\n",
      "            ROS1\n",
      "         SLC34A2\n",
      "   TRANSLOCATION\n",
      "Antisense fusion\n",
      "            17.0\n",
      "             6.0\n",
      "             yes\n",
      "         unknown\n",
      "\n"
     ]
    }
   ],
   "source": [
    "def format_msk_impact_report_with_variants(tables):\n",
    "    # Base clinical info from earlier\n",
    "\n",
    "    # Get variant tables\n",
    "    variant_tables = tables\n",
    "\n",
    "    # Build the report with variants\n",
    "    report_lines = [\"\\nDETECTED VARIANTS\\n\"]\n",
    "\n",
    "    if \"SNV/Indel\" in variant_tables:\n",
    "        report_lines.append(\"  Single Nucleotide Variants / Indels\")\n",
    "        report_lines.append(variant_tables[\"SNV/Indel\"].to_string(index=False))\n",
    "        report_lines.append(\"\")\n",
    "\n",
    "    if \"CNA\" in variant_tables:\n",
    "        report_lines.append(\"  Copy Number Alterations\")\n",
    "        report_lines.append(variant_tables[\"CNA\"].to_string(index=False))\n",
    "        report_lines.append(\"\")\n",
    "\n",
    "    if \"SV\" in variant_tables:\n",
    "        report_lines.append(\"  Structural Variants / Fusions\")\n",
    "        report_lines.append(variant_tables[\"SV\"].to_string(index=False))\n",
    "        report_lines.append(\"\")\n",
    "\n",
    "    return \"\\n\".join(report_lines)\n",
    "report_text_with_variants = format_msk_impact_report_with_variants(tables)\n",
    "print(report_text_with_variants)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe3a835d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
